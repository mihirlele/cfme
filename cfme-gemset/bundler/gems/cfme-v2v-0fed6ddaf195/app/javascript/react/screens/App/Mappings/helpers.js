export const addAssociatedPlansCount = mappings =>
  mappings.map(mapping => {
    mapping.associatedPlansCount = mapping.service_templates && mapping.service_templates.length;
    return mapping;
  });
