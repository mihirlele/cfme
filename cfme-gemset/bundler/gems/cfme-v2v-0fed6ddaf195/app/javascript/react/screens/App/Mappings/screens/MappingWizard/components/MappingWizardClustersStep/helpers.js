export const getProviderIds = clusters => {
  const providerIds = new Set();
  clusters.forEach(cluster => providerIds.add(cluster.ems_id));
  return [...providerIds];
};
