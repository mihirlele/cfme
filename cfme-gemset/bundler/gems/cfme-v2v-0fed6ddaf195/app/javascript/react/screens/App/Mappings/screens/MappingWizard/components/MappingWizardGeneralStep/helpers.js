export const asyncValidate = (values, dispatch, props) =>
  new Promise((resolve, reject) => {
    const { name: newTransformationName } = values;
    const { transformationMappings, editingMapping } = props;
    const existingTransformationNames = transformationMappings.reduce(
      (names, transformationMapping) => [...names, transformationMapping.name.trim()],
      []
    );
    const duplicateName = existingTransformationNames.find(
      existingTransformationName => existingTransformationName === newTransformationName.trim()
    );
    const duplicateIsEditingMappingName = editingMapping && duplicateName === editingMapping.name;

    if (duplicateName && !duplicateIsEditingMappingName) {
      props.showAlertAction(sprintf(__('Infrastructure mapping %s already exists'), newTransformationName));
      const error = { name: __('Mapping name already exists. Enter a unique name.') };
      reject(error);
    } else {
      resolve();
    }
  });

export const onChange = (values, dispatch, props) => {
  if (props.valid) {
    props.hideAlertAction();
  }
};
