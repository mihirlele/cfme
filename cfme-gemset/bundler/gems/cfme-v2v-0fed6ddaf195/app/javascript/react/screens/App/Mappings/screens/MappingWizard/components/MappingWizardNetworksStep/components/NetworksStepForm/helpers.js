import URI from 'urijs';

import {
  getTransformationMappingItemsBySourceType,
  getTransformationMappingItemsByDestinationType
} from '../../../../helpers';
import API from '../../../../../../../../../../common/API';
import {
  TRANSFORMATION_MAPPING_ITEM_SOURCE_TYPES,
  TRANSFORMATION_MAPPING_ITEM_DESTINATION_TYPES
} from '../../../../MappingWizardConstants';
import {
  QUERY_ATTRIBUTES,
  FETCH_NETWORK_URLS,
  NETWORK_ATTRIBUTES,
  HAS_CLOUD_NETWORKS
} from '../../MappingWizardNetworksStepConstants';
import { uniqueNetworks, combineNetworks } from '../../MappingWizardNetworksStepSelectors';
import { networkKey } from '../../../../../../../common/networkKey';

export const getRepresentatives = (groupedNetworks = {}) => {
  const representatives = [];
  for (const key in groupedNetworks) {
    if ({}.hasOwnProperty.call(groupedNetworks, key)) {
      representatives.push(groupedNetworks[key][0]);
    }
  }
  return representatives;
};

export const sourceNetworksFilter = (groupedSourceNetworks, networksStepMappings) => {
  const mappedNetworks = networksStepMappings.reduce((mappedNetworksArray, targetClusterWithNetworkMappings) => {
    const sourceNetworks = targetClusterWithNetworkMappings.nodes.reduce(
      (networks, networkMapping) => [...networks, ...networkMapping.nodes],
      []
    );
    return [...mappedNetworksArray, ...sourceNetworks];
  }, []);

  return getRepresentatives(groupedSourceNetworks).filter(
    network => !mappedNetworks.some(mappedNetwork => networkKey(mappedNetwork) === networkKey(network))
  );
};

export const clustersMappingWithTreeViewAttrs = clusterMapping => ({
  ...clusterMapping,
  text: clusterMapping.name,
  selectable: false
});

export const targetNetworkWithTreeViewAttrs = targetNetwork => ({
  ...targetNetwork,
  text: `${targetNetwork.providerName} \\ ${targetNetwork.name}`,
  selectable: true,
  selected: false,
  state: {
    expanded: true
  }
});

export const sourceNetworkWithTreeViewAttrs = (sourceNetwork, selectedCluster) => ({
  ...sourceNetwork,
  text: `${sourceNetwork.providerName} \\ ${sourceNetwork.name}`,
  icon: 'fa fa-file-o',
  sourceClusterId: selectedCluster.id,
  selectable: true,
  selected: false
});

export const networkGroupingForRep = (sourceNetworkRep, groupedSourceNetworks, selectedCluster) => {
  const sourceNetworks = groupedSourceNetworks[networkKey(sourceNetworkRep)];
  return sourceNetworks.map(sourceNetwork => sourceNetworkWithTreeViewAttrs(sourceNetwork, selectedCluster));
};

export const dedupeMappedSourceNetworks = networksMapping => {
  const { nodes: sourceNetworks, ...targetNetwork } = networksMapping;
  const groupedNetworks = uniqueNetworks(sourceNetworks);
  return {
    ...targetNetwork,
    nodes: getRepresentatives(groupedNetworks)
  };
};

export const dedupeMappedTargetNetworks = networksStepMapping => {
  const { nodes: networksMappings, ...targetCluster } = networksStepMapping;
  const groupedMappings = uniqueNetworks(networksMappings);
  return {
    ...targetCluster,
    nodes: getRepresentatives(groupedMappings)
  };
};

export const mappingsForTreeView = mappings =>
  mappings.reduce((updatedMappings, networksStepMapping) => {
    const { nodes: networksMappings } = dedupeMappedTargetNetworks(networksStepMapping);
    const updatedNetworksMappings = networksMappings.reduce(
      (dedupedMappings, networksMapping) => [...dedupedMappings, dedupeMappedSourceNetworks(networksMapping)],
      []
    );
    const updatedMapping = {
      ...networksStepMapping,
      nodes: updatedNetworksMappings
    };
    return [...updatedMappings, updatedMapping];
  }, []);

export const mappingWithTargetNetworkRemoved = (networksStepMapping, targetNetworkToRemove) => {
  const { nodes: networksMappings, ...targetCluster } = networksStepMapping;
  const updatedNetworksMappings = networksMappings.filter(
    targetNetworkWithSourceNetworks => targetNetworkWithSourceNetworks.id !== targetNetworkToRemove.id
  );
  return updatedNetworksMappings.length === 0
    ? null
    : {
        ...targetCluster,
        nodes: updatedNetworksMappings
      };
};

export const mappingWithSourceNetworkRemoved = (networksMapping, sourceNetworkToRemove) => {
  const { nodes: sourceNetworks, ...targetNetwork } = networksMapping;
  const updatedSourceNetworks = sourceNetworks.filter(
    sourceNetwork => networkKey(sourceNetwork) !== networkKey(sourceNetworkToRemove)
  );
  return updatedSourceNetworks.length === 0
    ? null
    : {
        ...targetNetwork,
        nodes: updatedSourceNetworks
      };
};

const getMappedTargetNetworks = (networks, mappedNetworkIds) =>
  networks.filter(network => mappedNetworkIds.some(id => id === network.id));

const getMappedSourceNetworks = (targetNetwork, sourceNetworks, transformation) => {
  const networkMappingItems = getTransformationMappingItemsBySourceType(
    TRANSFORMATION_MAPPING_ITEM_SOURCE_TYPES.network,
    transformation
  );
  return sourceNetworks.filter(sourceNetwork =>
    networkMappingItems.some(item => item.destination_id === targetNetwork.id && item.source_id === sourceNetwork.id)
  );
};

const clusterWithNetworksUrl = (id, provider) => {
  const uri = new URI(`${FETCH_NETWORK_URLS[provider]}/${id}`);
  uri.addSearch({ attributes: QUERY_ATTRIBUTES[provider] });

  return uri.toString();
};

const targetNetworkPromises = (targetClusterIds, targetProvider) =>
  targetClusterIds.map(
    id =>
      new Promise((resolve, reject) => {
        API.get(clusterWithNetworksUrl(id, targetProvider))
          .then(res => {
            const updatedTargetNetworks = res.data[NETWORK_ATTRIBUTES[targetProvider]].map(targetNetwork => ({
              ...targetNetwork,
              clusterId: id,
              providerName: res.data.ext_management_system.name
            }));
            resolve({ ...res.data, [NETWORK_ATTRIBUTES[targetProvider]]: updatedTargetNetworks });
          })
          .catch(e => reject(e));
      })
  );

const sourceNetworkPromises = sourceClusterIds =>
  sourceClusterIds.map(
    id =>
      new Promise((resolve, reject) => {
        API.get(clusterWithNetworksUrl(id, 'source'))
          .then(res => {
            resolve(
              res.data.lans.map(lan => ({
                ...lan,
                sourceClusterId: id,
                clusterId: id,
                providerName: res.data.ext_management_system.name
              }))
            );
          })
          .catch(e => reject(e));
      })
  );

const publicNetworksUrl = targetProviderId => {
  const uri = new URI(`${FETCH_NETWORK_URLS.public}/${targetProviderId}/cloud_networks`);
  uri.addSearch({ expand: 'resources' });
  return uri.toString();
};

const publicNetworksPromises = (clusterMappings, targetProvider) => {
  if (!HAS_CLOUD_NETWORKS[targetProvider]) return [Promise.resolve([])];
  return clusterMappings.map(
    mapping =>
      new Promise((resolve, reject) => {
        API.get(publicNetworksUrl(mapping.ext_management_system.id))
          .then(res =>
            resolve(
              res.data.resources.map(network => ({
                ...network,
                providerName: mapping.ext_management_system.name,
                clusterId: mapping.id
              }))
            )
          )
          .catch(e => reject(e));
      })
  );
};

export const createNetworksMappings = (transformation, targetProvider, clusterMappings) =>
  new Promise((resolve, reject) => {
    const targetClusterIds = getTransformationMappingItemsByDestinationType(
      TRANSFORMATION_MAPPING_ITEM_DESTINATION_TYPES[targetProvider].cluster,
      transformation
    ).map(item => item.destination_id);
    const sourceClusterIds = getTransformationMappingItemsBySourceType(
      TRANSFORMATION_MAPPING_ITEM_SOURCE_TYPES.cluster,
      transformation
    ).map(item => item.source_id);
    const targetNetworkIds = getTransformationMappingItemsByDestinationType(
      TRANSFORMATION_MAPPING_ITEM_DESTINATION_TYPES[targetProvider].network,
      transformation
    ).map(item => item.destination_id);

    Promise.all([
      Promise.all(targetNetworkPromises(targetClusterIds, targetProvider)),
      Promise.all(sourceNetworkPromises(sourceClusterIds)),
      Promise.all(publicNetworksPromises(clusterMappings, targetProvider))
    ])
      .then(res => {
        const networksMappings = [];
        const [targetClustersWithNetworks, sourceNetworkResults, publicNetworksResults] = res;
        const sourceNetworks = sourceNetworkResults.reduce(
          (networks, resolvedPromise) => [...networks, ...resolvedPromise],
          []
        );
        const publicNetworks = publicNetworksResults.reduce(
          (networks, resolvedPromise) => [...networks, ...resolvedPromise],
          []
        );
        targetClustersWithNetworks.forEach(mapping => {
          const privateNetworks = mapping[NETWORK_ATTRIBUTES[targetProvider]];
          const nodes = getMappedTargetNetworks(combineNetworks(privateNetworks, publicNetworks), targetNetworkIds).map(
            targetNetwork => ({
              ...targetNetworkWithTreeViewAttrs(targetNetwork),
              nodes: getMappedSourceNetworks(targetNetwork, sourceNetworks, transformation).map(network =>
                sourceNetworkWithTreeViewAttrs(network, { id: network.sourceClusterId })
              )
            })
          );

          networksMappings.push({
            ...mapping,
            text: mapping.name,
            selectable: false,
            state: {
              expanded: true
            },
            nodes
          });
        });

        resolve(networksMappings);
      })
      .catch(e => reject(e));
  });
