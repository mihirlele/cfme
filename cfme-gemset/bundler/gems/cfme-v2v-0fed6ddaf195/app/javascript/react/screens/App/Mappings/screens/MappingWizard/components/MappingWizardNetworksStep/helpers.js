export const normalizeCloudNetworkHref = network => {
  const normalizedHref = network.href && network.href.replace(/\/providers\/\d+/, '');

  return { ...network, href: normalizedHref };
};
