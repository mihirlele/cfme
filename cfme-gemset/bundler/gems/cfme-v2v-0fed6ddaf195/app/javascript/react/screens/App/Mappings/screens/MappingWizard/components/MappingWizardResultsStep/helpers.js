import { SOURCE_HREF_SLUGS, DESTINATION_HREF_SLUGS } from './MappingWizardResultsStepConstants';

const normalizeTransformationItem = (transformationItem, targetProvider) => ({
  source: `${SOURCE_HREF_SLUGS[transformationItem.source_type]}${transformationItem.source_id}`,
  destination: `${DESTINATION_HREF_SLUGS[targetProvider][transformationItem.destination_type]}${
    transformationItem.destination_id
  }`
});

export const transformationHasBeenEdited = (transformation, postBody, targetProvider) => {
  const originalTransformationItems = JSON.stringify(
    transformation.transformation_mapping_items.map(item => normalizeTransformationItem(item, targetProvider))
  );
  const currentTransformationItems = JSON.stringify(postBody.transformation_mapping_items);

  return (
    transformation.name !== postBody.name ||
    transformation.description !== postBody.description ||
    originalTransformationItems !== currentTransformationItems
  );
};
