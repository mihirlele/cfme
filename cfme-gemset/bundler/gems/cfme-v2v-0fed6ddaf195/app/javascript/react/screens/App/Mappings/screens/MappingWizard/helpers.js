import { TRANSFORMATION_MAPPING_ITEM_DESTINATION_TYPES } from './MappingWizardConstants';

export const mappingUrlRegex = /\/api\/\w{1,}\/\d{1,}/;

export const createTransformationMappings = (
  mappingWizardGeneralStep,
  mappingWizardClustersStep,
  mappingWizardDatastoresStep,
  mappingWizardNetworksStep
) => {
  const clusterTransformationMappings = mappingWizardClustersStep.values.clusterMappings.reduce(
    (clusterTransformationsArray, targetClusterWithSourceClusters) => {
      const destination = targetClusterWithSourceClusters.href.match(mappingUrlRegex)[0];
      const transformations = targetClusterWithSourceClusters.nodes.reduce(
        (clusterTransformations, sourceCluster) =>
          clusterTransformations.concat({
            source: sourceCluster.href.match(mappingUrlRegex)[0],
            destination
          }),
        []
      );
      return clusterTransformationsArray.concat(transformations);
    },
    []
  );

  const datastoreTransformationMappings = mappingWizardDatastoresStep.values.datastoresMappings.reduce(
    (datastoreTransformationsPerTargetCluster, targetClusterWithDatastoreMappings) => {
      const datastoreTransformationsForTargetCluster = targetClusterWithDatastoreMappings.nodes.reduce(
        (datastoreTransformationsPerDatastoreMapping, targetDatastoreWithSourceDatastores) => {
          const datastoreTransformations = targetDatastoreWithSourceDatastores.nodes.reduce(
            (transformations, sourceDatastore) =>
              transformations.concat({
                source: sourceDatastore.href.match(mappingUrlRegex)[0],
                destination: targetDatastoreWithSourceDatastores.href.match(mappingUrlRegex)[0]
              }),
            []
          );
          return datastoreTransformationsPerDatastoreMapping.concat(datastoreTransformations);
        },
        []
      );
      return datastoreTransformationsPerTargetCluster.concat(datastoreTransformationsForTargetCluster);
    },
    []
  );

  const networkTransformationMappings = mappingWizardNetworksStep.values.networksMappings.reduce(
    (networkTransformationsPerTargetCluster, targetClusterWithNetworkMappings) => {
      const networkTransformationsForTargetCluster = targetClusterWithNetworkMappings.nodes.reduce(
        (networkTransformationsPerNetworkMapping, targetNetworkWithSourceNetworks) => {
          const networkTransformations = targetNetworkWithSourceNetworks.nodes.reduce(
            (transformations, sourceNetwork) =>
              transformations.concat({
                source: sourceNetwork.href.match(mappingUrlRegex)[0],
                destination: targetNetworkWithSourceNetworks.href.match(mappingUrlRegex)[0]
              }),
            []
          );
          return networkTransformationsPerNetworkMapping.concat(networkTransformations);
        },
        []
      );
      return networkTransformationsPerTargetCluster.concat(networkTransformationsForTargetCluster);
    },
    []
  );

  return {
    name: mappingWizardGeneralStep.values.name,
    description: mappingWizardGeneralStep.values.description,
    state: 'draft',
    transformation_mapping_items: [
      ...clusterTransformationMappings,
      ...datastoreTransformationMappings,
      ...networkTransformationMappings
    ]
  };
};

export const getMappedSourceClusters = clusterMappings =>
  clusterMappings.reduce(
    (sourceClusters, targetClusterWithSourceClusters) => sourceClusters.concat(targetClusterWithSourceClusters.nodes),
    []
  );

export const getSourceClustersWithMappings = mappings =>
  mappings.reduce((idsPerTargetCluster, targetClusterWithMappings) => {
    const idsForTargetCluster = targetClusterWithMappings.nodes.reduce(
      (idsPerMapping, targetObjectWithSourceObjects) => {
        const idsForSourceObjects = targetObjectWithSourceObjects.nodes.map(
          sourceObject => sourceObject.sourceClusterId
        );
        return idsPerMapping.concat(Array.from(new Set(idsForSourceObjects)));
      },
      []
    );
    return idsPerTargetCluster.concat(Array.from(new Set(idsForTargetCluster)));
  }, []);

export const groupClusterTransformationItemsByDestinationId = (transformationItems = []) =>
  transformationItems.reduce(
    (map, item) => ({
      ...map,
      [item.destination_id]: map[item.destination_id] ? [...map[item.destination_id], item.source_id] : [item.source_id]
    }),
    {}
  );

export const getTransformationMappingItemsBySourceType = (type, transformation) =>
  transformation.transformation_mapping_items.filter(item => item.source_type === type);

export const getTransformationMappingItemsByDestinationType = (type, transformation) =>
  transformation.transformation_mapping_items.filter(item => item.destination_type === type);

export const determineTargetProvider = transformation =>
  getTransformationMappingItemsByDestinationType(
    TRANSFORMATION_MAPPING_ITEM_DESTINATION_TYPES.openstack.cluster,
    transformation
  ).length
    ? 'openstack'
    : 'rhevm';
