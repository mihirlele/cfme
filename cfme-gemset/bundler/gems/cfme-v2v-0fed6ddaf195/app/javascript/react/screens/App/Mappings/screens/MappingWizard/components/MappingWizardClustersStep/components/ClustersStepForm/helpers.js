import { groupClusterTransformationItemsByDestinationId } from '../../../../helpers';

export const targetClusterWithExtendedData = targetCluster => ({
  ...targetCluster,
  text: targetCluster.name,
  state: {
    expanded: true
  },
  selectable: true,
  selected: false
});

export const sourceClusterWithExtendedData = sourceCluster => ({
  ...sourceCluster,
  text: sourceCluster.name,
  icon: 'fa fa-file-o'
});

export const updateMapping = (clustersStepMapping, targetClusterToAddTo, sourceClustersToAdd) => {
  const { nodes: sourceClusters, ...targetCluster } = clustersStepMapping;

  if (targetCluster.id === targetClusterToAddTo.id) {
    return {
      ...targetCluster,
      nodes: [
        ...sourceClusters,
        ...sourceClustersToAdd.map(sourceCluster => sourceClusterWithExtendedData(sourceCluster))
      ]
    };
  }
  return clustersStepMapping;
};

export const createNewMapping = (targetCluster, sourceClusters) => ({
  ...targetClusterWithExtendedData(targetCluster),
  nodes: sourceClusters.map(sourceCluster => sourceClusterWithExtendedData(sourceCluster))
});

export const createClusterMappings = (transformationItems, targetClusters, sourceClusters) => {
  const mappings = groupClusterTransformationItemsByDestinationId(transformationItems);

  if (targetClusters.length && sourceClusters.length) {
    return Object.keys(mappings).map(key => {
      const target = targetClusters && targetClusters.find(cluster => cluster.id === key);
      const sources = sourceClusters && sourceClusters.filter(cluster => mappings[key].some(id => id === cluster.id));

      return createNewMapping(target, sources);
    });
  }
  return [];
};

export const providerHasSshKeyPair = (target, providers) => {
  const provider = providers.find(prov => prov.id === target.ems_id);

  return (
    provider.authentications &&
    provider.authentications.some(authentication => authentication.authtype === 'ssh_keypair')
  );
};
