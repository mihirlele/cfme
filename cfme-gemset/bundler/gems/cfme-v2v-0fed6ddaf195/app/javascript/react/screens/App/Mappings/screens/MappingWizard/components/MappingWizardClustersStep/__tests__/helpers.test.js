import { getProviderIds } from '../helpers';

describe('getProviderIds', () => {
  test('returns an array with unique ems IDs', () => {
    const clusters = [
      {
        ems_id: '1'
      },
      {
        ems_id: '2'
      },
      {
        ems_id: '1'
      }
    ];

    expect(getProviderIds(clusters)).toMatchSnapshot();
  });
});
