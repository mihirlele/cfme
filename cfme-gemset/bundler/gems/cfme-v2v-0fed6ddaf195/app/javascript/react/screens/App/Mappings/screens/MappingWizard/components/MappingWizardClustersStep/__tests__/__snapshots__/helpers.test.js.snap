// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getProviderIds returns an array with unique ems IDs 1`] = `
Array [
  "1",
  "2",
]
`;
